src/RustyObjectStore.jl

```diff
@@ -82,17 +82,61 @@ const DEFAULT_CONFIG = StaticConfig(
     concurrency_limit=512
 )
 
+function default_panic_hook()
+    println("Rust thread panicked, exiting the process")
+    exit(1)
+end
+
 const _OBJECT_STORE_STARTED = Ref(false)
 const _INIT_LOCK::ReentrantLock = ReentrantLock()
+_PANIC_HOOK::Function = default_panic_hook
 
 struct InitException <: Exception
     msg::String
     return_code::Cint
 end
 
-function default_panic_hook()
-    println("Rust thread panicked, exiting the process")
-    exit(1)
+Base.@ccallable function panic_hook_wrapper()::Cint
+    global _PANIC_HOOK
+    _PANIC_HOOK()
+    return 0
+end
+
+# This is the callback that Rust calls to notify a Julia task of a completed operation.
+# The argument is transparent to Rust and is simply what gets passed from Julia in the handle
+# argument of the @ccall. Currently we pass a pointer to a Base.Event that must be notified to
+# wakeup the appropriate task.
+Base.@ccallable function notify_result(event_ptr::Ptr{Nothing})::Cint
+    event = unsafe_pointer_to_objref(event_ptr)
```

Comment:
Yeah, I think it must be dynamic dispatch due to `Ptr{Nothing}` if you think having the type constraint makes Julia trust me that the pointer has the correct type that would be nice

Reply:
Actually it does not look like dynamic dispatch:
```
@code_llvm unsafe_pointer_to_objref(ptr)
;  @ pointer.jl:253 within `unsafe_pointer_to_objref`
define nonnull {}* @julia_unsafe_pointer_to_objref_330(i64 zeroext %0) #0 {
top:
  %1 = inttoptr i64 %0 to {}*
  ret {}* %1
}
```
```
@code_native unsafe_pointer_to_objref(ptr)
	.text
	.file	"unsafe_pointer_to_objref"
	.globl	julia_unsafe_pointer_to_objref_114 # -- Begin function julia_unsafe_pointer_to_objref_114
	.p2align	4, 0x90
	.type	julia_unsafe_pointer_to_objref_114,@function
julia_unsafe_pointer_to_objref_114:     # @julia_unsafe_pointer_to_objref_114
; ┌ @ pointer.jl:253 within `unsafe_pointer_to_objref`
# %bb.0:                                # %top
	push	rbp
	mov	rbp, rsp
	mov	rax, rdi
	pop	rbp
	ret
.Lfunc_end0:
	.size	julia_unsafe_pointer_to_objref_114, .Lfunc_end0-julia_unsafe_pointer_to_objref_114
; └
                                        # -- End function
	.section	".note.GNU-stack","",@progbits
```

Reply:
This tells otherwise:
```
@code_lowered unsafe_pointer_to_objref(ptr)
CodeInfo(
1 ─ %1 = Core.apply_type(Base.Ptr, Base.Cvoid)
│   %2 = Base.cconvert(%1, x)
│   %3 = Core.apply_type(Base.Ptr, Base.Cvoid)
│   %4 = Base.unsafe_convert(%3, %2)
│   %5 = $(Expr(:foreigncall, :(:jl_value_ptr), Any, svec(Ptr{Nothing}), 0, :(:ccall), :(%4), :(%2)))
└──      return %5
)
```

Reply:
Yes,  I think this is a good idea:
![image](https://github.com/RelationalAI/RustyObjectStore.jl/assets/2074313/64f6a84e-81a8-4a08-8637-8d4a72782888)
vs
![image](https://github.com/RelationalAI/RustyObjectStore.jl/assets/2074313/603ece82-42b4-4a11-9fb4-6d4c0027e00e)